src/edu/upenn/cit594/datamanagement/DelimitedFileReader.java

```diff
@@ -0,0 +1,48 @@
+package edu.upenn.cit594.datamanagement;
+
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.*;
+
+public abstract class DelimitedFileReader<T> extends FileReader<T> {
+    boolean hasHeaders;
+    String delimitBy;
+
+    public DelimitedFileReader(String fileName, boolean hasHeaders, String delimitBy) {
+        super(fileName);
+        this.hasHeaders = hasHeaders;
+        this.delimitBy = delimitBy;
+    }
+
+    public T read() {
+        String line = null;
+        List<String> headerList = null;
+
+        try {
+            BufferedReader br = new BufferedReader(new java.io.FileReader(this.fileName));
+
+            if(this.hasHeaders) {
+                line = br.readLine();
+                headerList = Arrays.asList(line.split(this.delimitBy));
+
+                setHeaderIndices(headerList);
+            }
+
+            while((line = br.readLine()) != null) {
+                List<String> dataList = Arrays.asList(line.split(this.delimitBy));
```

Comment:
I think this could be separated into a tokenize method (tokenize(String line)) that deals with the logic of splitting. Would work on the headerList line (line 27) as well.

Reply:
`/**
     * Splits line into list of string tokens based on inlineDelimiter.
     * @param line the line to split
     * @return the tokens that were found
     */
    protected List<String> tokenize(String line) {
        if (line == null) return null;
        List<String> output = new ArrayList<>();
        Matcher matcher = matchPattern.matcher(line);
        // find all matches in provided line that correspond to matchPattern and add to output list
        while (matcher.find()) {
            String result = matcher.group();
            // remove any leading or trailing quotes, inlineDelimiter and whitespace
            if (result.startsWith("\"")) { result = result.substring(1); }
            if (result.endsWith(inlineDelimiter)) { result = result.substring(0, result.length() - 1); }
            if (result.endsWith("\"")) { result = result.substring(0, result.length() - 1); }
            
            output.add(result.trim());
        }
        return output;
    }`

Reply:
matchPattern = Pattern.compile( // matches with...
                        "\"[^\"]*\"" + inlineDelimiter // double quotes followed by the inlineDelimiter ("Shapiro, Eric", => "Shapiro, Eric",)
                                + "|" + "[^"+ inlineDelimiter + "]+" // anything that isn't the inlineDelimiter (test, => test)
                                + "|" + "^" + inlineDelimiter // a starting inlineDelimiter (, => ,) this is for when start of line has no value
                                + "|" + "(?=" + inlineDelimiter + "(" + inlineDelimiter + "|$))" // the inlineDelimiter followed by the inlineDelimiter or the end of the line (,, => "")
        );